loader/windows/icd_windows_dxgk.c

```diff
@@ -0,0 +1,141 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * OpenCL is a trademark of Apple Inc. used under license by Khronos.
+ */
+
+#include "icd.h"
+#include <windows.h>
+#include "icd_windows_dxgk.h"
+#include <assert.h>
+
+#ifndef NTSTATUS
+typedef LONG NTSTATUS;
+#define STATUS_SUCCESS  ((NTSTATUS)0x00000000L)
+#define STATUS_BUFFER_TOO_SMALL ((NTSTATUS)0xC0000023)
+#define NT_SUCCESS(Status) (((NTSTATUS)(Status)) >= 0)
+#endif
+
+#include "d3dkmthk.h"
+
+bool khrIcdOsVendorsEnumerateDXGK(void)
+{
+    bool ret = FALSE;
+#if defined(DXGKDDI_INTERFACE_VERSION_WDDM2_4) && (DXGKDDI_INTERFACE_VERSION >= DXGKDDI_INTERFACE_VERSION_WDDM2_4)
+    {
+        D3DKMT_ADAPTERINFO* pAdapterInfo = NULL;
+        D3DKMT_ENUMADAPTERS2 EnumAdapters;
+        NTSTATUS Status = STATUS_SUCCESS;
+
+        // Get handle to GDI Runtime
+        HMODULE h = LoadLibrary("gdi32.dll");
+        KHR_ICD_ASSERT(h != NULL);
+
+        char cszLibraryName[MAX_PATH] = { 0 };
+        PFND3DKMT_ENUMADAPTERS2 pEnumAdapters2 = (PFND3DKMT_ENUMADAPTERS2)GetProcAddress((HMODULE)h, "D3DKMTEnumAdapters2");
+        if (!pEnumAdapters2)
+        {
+            KHR_ICD_TRACE("GetProcAddress failed for D3DKMT_ENUMADAPTERS2\n");
+            goto out;
+        }
+        while (1)
+        {
+            EnumAdapters.NumAdapters = 0;
+            EnumAdapters.pAdapters = NULL;
+            Status = pEnumAdapters2(&EnumAdapters);
+            if (Status == STATUS_BUFFER_TOO_SMALL)
+            {
+                // Number of Adapters increased between calls, retry;
+                continue;
+            }
+            else if (!NT_SUCCESS(Status))
+            {
+                KHR_ICD_TRACE("D3DKMT_ENUMADAPTERS2 status != SUCCESS\n");
+                goto out;
+            }
+            break;
+        }
+        pAdapterInfo = (D3DKMT_ADAPTERINFO*)malloc(sizeof(D3DKMT_ADAPTERINFO)*(EnumAdapters.NumAdapters));
+        if (pAdapterInfo == NULL)
+        {
+            KHR_ICD_TRACE("Allocation failure for AdapterInfo buffer\n");
+            goto out;
+        }
+        EnumAdapters.pAdapters = pAdapterInfo;
+        Status = pEnumAdapters2(&EnumAdapters);
+        if (!NT_SUCCESS(Status))
+        {
+            KHR_ICD_TRACE("D3DKMT_ENUMADAPTERS2 status != SUCCESS\n");
+            goto out;
+        }
+        for (UINT AdapterIndex = 0; AdapterIndex < EnumAdapters.NumAdapters; AdapterIndex++)
+        {
+            D3DDDI_QUERYREGISTRY_INFO QueryArgs = {0};
+            D3DDDI_QUERYREGISTRY_INFO* pQueryArgs = &QueryArgs;
+            D3DDDI_QUERYREGISTRY_INFO* pQueryBuffer = NULL;
+            QueryArgs.QueryType = D3DDDI_QUERYREGISTRY_ADAPTERKEY;
+            QueryArgs.QueryFlags.TranslatePath = TRUE;
+            QueryArgs.ValueType = REG_MULTI_SZ;
```

Comment:
Here is the check for `REG_MULTI_SZ`.  The ICD spec currently says that only `REG_SZ` registry values are searched.  Is there some reason why this code is using `REG_MULTI_SZ`?

Reply:
Just fixed it. Sorry I've accumulated various changes from our trunk, and majority of the confusion here is because OpenGL string is MULTI_SZ so our installer was creating that incorrectly.